lib/src/authentication/token_store.dart

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import 'dart:convert';
+
+import 'package:path/path.dart' as path;
+
+import '../io.dart';
+import '../log.dart' as log;
+import '../source/hosted.dart';
+import 'token.dart';
+
+/// Stores and manages authentication credentials.
+class TokenStore {
+  TokenStore(this.cacheRootDir);
+
+  /// Cache directory.
+  final String cacheRootDir;
+
+  List<Token>? _tokens;
+
+  /// List of saved authentication tokens.
+  ///
+  /// Modifying this field will not write changes to the disk. You have to call
+  /// [flush] to save changes.
+  List<Token> get tokens => _tokens ??= _loadTokens();
+
+  /// Reads "tokens.json" and parses / deserializes it into list of
+  /// [Token].
+  List<Token> _loadTokens() {
+    final result = List<Token>.empty(growable: true);
+    final path = _tokensFile;
+    if (!fileExists(path)) {
+      return result;
+    }
+
+    try {
+      final json = jsonDecode(readTextFile(path));
+
+      if (json is! Map<String, dynamic>) {
+        throw FormatException('JSON contents is corrupted or not supported.');
+      }
+      if (json['version'] != 1) {
+        throw FormatException('Version is not supported.');
+      }
+
+      if (json.containsKey('hosted')) {
+        if (json['hosted'] is! List) {
+          throw FormatException(
+              'tokens.json format is invalid or not supported.');
+        }
+
+        result.addAll((json['hosted'] as List)
+            .cast<Map<String, dynamic>>()
+            .map((it) => Token.fromJson(it)));
+      }
+    } on FormatException catch (error, stackTrace) {
+      log.error('Failed to load tokens.json', error, stackTrace);
+
+      // When an invalid, damaged or not compatible version of token.json is
+      // found, we remove it after showing error message. Otherwise the error
+      // message will be displayed on each pub command.
+      // Or instead we could write instructrions to execute
+      //`pub token remove --all` if user couldn't solve the issue.
+      deleteEntry(_tokensFile);
+    }
+
+    return result;
+  }
+
+  /// Writes [tokens] into "tokens.json".
+  void _saveTokens(List<Token> tokens) {
+    writeTextFile(
+        _tokensFile,
+        jsonEncode(<String, dynamic>{
+          'version': 1,
+          'hosted': tokens.map((it) => it.toJson()).toList(),
```

Comment:
Which way would you suggest?

A. Opening & json de-serializing the file again and modifying `hosted` field.
B. Caching json contents on `_loadTokens` and using cached value to modify `hosted` field.

Reply:
I think either solution would work. I think the only concern is race conditions, when the user is running two pub commands concurrently.

Perhaps (A) is best: when modifying `tokens.json` we just load it, do the modification and save it back (retaining unknown fields on objects that are not modified).

Example:
```js
// If modifying token for "https://some-pub-server.com", then
{
  "version": 1,
  "hosted": [
    {
      "url": "https://some-pub-server.com"
      "token": "..."
      "unknownField": ..., // delete this unknown field because we're modifying this object
    },
    {
      "url": "https://other-pub-server.com"
      "token": "..."
      "unknownFieldB": ...,  // preserve this unknown field because we're not modifying this object.
    },
  ],
}
```

Doesn't that sound like behavior that will allow us to be reasonably future proof?

Reply:
It's also possible that what I proposed is too complicated.

I think what I'm suggesting is that we think about how new features can be added to `tokens.json`, and how old clients should behave when this happens.

Ideally, we could leave a comment in here outlining:
 * What old clients should do when seeing a new feature,
 * How new clients can introduce new features without breaking old clients horribly.

This is worth thinking about because users sometimes switches back and forth between two different SDK versions. And if we don't consider upgrade-strategy upfront we might have painted ourselves into a corner where it's hard to add new features.